python/ray/data/datasource/spark_datasource.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import os
+from typing import Iterator, List, Optional
+import numpy as np
+import pyarrow
+
+from ray.data.block import BlockMetadata
+from ray.data.datasource.datasource import Datasource, ReadTask
+from ray.util.annotations import PublicAPI
+
+
+def validate_requirements():
+    from ray.util.spark.utils import get_spark_session
+
+    spark = get_spark_session()
+    if spark.conf.get(
+        "spark.databricks.pyspark.dataFrameChunk.enabled", "false"
+    ).lower() != "true":
+        raise RuntimeError(
+            "In databricks runtime, if you want to use 'ray.data.from_spark' API, "
+            "you need to set spark cluster config "
+            "'spark.databricks.pyspark.dataFrameChunk.enabled' to 'true'."
+        )
+
+    if spark.conf.get(
+        "spark.databricks.acl.dfAclsEnabled", "false"
+    ).lower() != "false":
+        raise RuntimeError(
+            "In databricks runtime, if you want to use 'ray.data.from_spark' API, "
+            "you must use an assigned mode databricks cluster."
+        )
+
+
+def _persist_dataframe_as_chunks(spark_dataframe, bytes_per_chunk):
+    try:
+        from pyspark.databricks.sql.chunk import persistDataFrameAsChunks
+    except ImportError:
+        raise RuntimeError(
+            "Current Databricks Runtime version does not support Ray SparkDatasource."
+        )
+    return persistDataFrameAsChunks(spark_dataframe, bytes_per_chunk)
+
+
+def _read_chunk_fn(chunk_ids) -> Iterator["pyarrow.Table"]:
+    if read_chunk_fn_path := os.environ.get(
+        "_RAY_DATABRICKS_FROM_SPARK_READ_CHUNK_FN_PATH"
+    ):
+        import ray.cloudpickle as pickle
+
+        # This is for testing.
+        with open(read_chunk_fn_path, "rb") as f:
+            readChunk = pickle.load(f)
+
+    else:
+        from pyspark.databricks.sql.chunk import readChunk
+
+    yield from map(readChunk, chunk_ids)
```

Comment:
Can we unpersist the chunk on-the-fly after reading the chunk?

Reply:
This sounds not good, because Ray task might be retried if some error occurs, if we unpersist it, the retrying task will fail.

On the other hand, we don't cache the chunk data in Ray side, then it meaning if we traverse the Ray dataset a second time (this might be common cases), it requires to read from spark cache again.